skt/__init__.py

```diff
@@ -476,16 +477,38 @@ def __init__(self, path, basecfg, cfgtype=None, makeopts=None,
         self.basecfg = os.path.expanduser(basecfg)
         self.cfgtype = cfgtype if cfgtype is not None else "olddefconfig"
         self._ready = 0
-        self.makeopts = None
         self.buildlog = "%s/build.log" % self.path
         self.defmakeargs = ["make", "-C", self.path]
         self.enable_debuginfo = enable_debuginfo
 
         if makeopts is not None:
-            # FIXME: Might want something a bit smarter here, something that
-            # would parse it the same way bash does
-            self.makeopts = makeopts.split(' ')
-            self.defmakeargs += self.makeopts
+            self.makeopts = self.defmakeargs + shlex.split(makeopts)
+        else:
+            self.makeopts = self.defmakeargs
+
+        # If the user provided a directory to hold kernel build artifacts,
+        # we must add the appropriate make options.
+        if kbuild_output is not None:
+            # Ensure the path is absolute, even if the user provided a path
+            # relative to their home directory with a tilde (~).
+            self.kbuild_output_dir = os.path.abspath(
+                os.path.expanduser(kbuild_output)
+            )
+            logging.info(
+                "Found kernel build output directory specifier in make "
+                "options: {}".format(self.kbuild_output_dir)
+            )
+
+            # Ensure the output directory exists
+            # NOTE(mhayden): os.makedirs will cause an error if the directory
+            # already exists, so we must check for the directory first.
+            if not os.path.isdir(self.kbuild_output_dir):
+                os.makedirs(self.kbuild_output_dir)
+
+            # Add the 'O=' make option to specify the output directory
+            self.makeopts.append("O={}".format(self.kbuild_output_dir))
+        else:
+            self.kbuild_output_dir = None
```

Comment:
Should we default this to self.path to avoid more if/else clauses below?

Reply:
Well, in the single work directory model, `self.path` points to a work directory that contains the kernel source and eventually the kernel build output. However, with the kernel build output separate from the source, you end up with:

* `self.path`: Contains kernel source
* `self.kbuild_output_dir`: Contains build output

Setting `self.path` to the kernel build output directory would break things since skt expects `self.path` to contain the kernel source repository.

Reply:
You have it backwards.
else:
    self.kbuild_output_dir = self.path

:-)  Look at your various if/else below, they all have if self.kbuild_output_dir then <use it> else <use self.path>.  By assigning self.kbuild_output_dir to self.path in the absence of --kbuild-output, simplifies the code and makes it easier to read. :-)